### ISSUE TITLE
[ttnn][conv2d] DRAM Bandwidth Bottleneck in Pointwise Conv2d with in_channels=3, kernel=1×1 on Wormhole N150

### ISSUE THREAD

**Issue by chandrasekaranpradeep** [tenstorrent/tt-metal#46831]
## Problem

`ttnn.conv2d` with `in_channels=3`, `out_channels=3`, `kernel_size=(1,1)`, and the activation
tensor in DRAM (`config_tensors_in_dram=True`) is severely DRAM-bandwidth bound on Wormhole N150.

Internally, `ttnn.conv2d` with `kernel_size=(1,1)` is executed as a **matrix multiplication**
(`MatmulDeviceOperation`). The activation tensor `[N, 1, H×W, C]` (where N=batch=1, the
inner `1` comes from the reshape, H×W are the spatial positions, and C=in_channels=3) is
treated as a matrix with **N×H×W rows** and **C columns**. Because `ttnn.conv2d` requires
NHWC layout, a `permute` is inserted before the call to reorder the input from NCHW. This
permute moves `C=3` to the tile column dimension. Since Wormhole tiles are always **32×32**,
the column is padded from 3 to 32 — only 3 out of 32 column slots in every tile hold real
data, 29 are zeros.

The physical activation size inflates by **10.67×**:
- Config 1 `[1,3,1536,1536]`: 14.2 MB real → 151.0 MB physical — 2,304 KB/core, 1.7× over L1 → DRAM
- Config 2 `[1,3,1280,2304]`: 17.7 MB real → 188.7 MB physical — 2,880 KB/core, 2.1× over L1 → DRAM

The `MatmulDeviceOperation` kernel reads **32 values** per row from DRAM, uses **3** for
computation, and discards the remaining **29 zero-padded values** — saturating the DRAM bus
while keeping the FPU at 0.008%.

Tracy profiling of the isolated conv2d kernel (`test_conv2d_only`, activation pre-placed in
DRAM in NHWC-flat layout) confirms:

| Metric | conv2d_1 `[1,3,1536,1536]` | conv2d_2 `[1,3,1280,2304]` |
|--------|:--------------------------:|:--------------------------:|
| Internal op | `MatmulDeviceOperation` | `MatmulDeviceOperation` |
| Device FW duration | **6.431 ms** | **7.924 ms** |
| Device kernel duration | 6.430 ms | 7.923 ms |
| PM IDEAL | 545,058 ns | 681,322 ns |
| PM BANDWIDTH = PM IDEAL | **545,058 ns** | **681,322 ns** |
| PM COMPUTE | 486 ns (0.09% of ideal) | 608 ns (0.09% of ideal) |
| **PM FPU UTIL** | **0.008%** | **0.008%** |
| **PM REQ INPUT BW** | **277.03 GB/s** | **277.03 GB/s** |
| Input W_PAD\[logical\] (N) | `1[1]` | `1[1]` |
| Input Z_PAD\[logical\] | `1[1]` | `1[1]` |
| Input Y_PAD\[logical\] (N×H×W) | `2359296[2359296]` | `2949120[2949120]` |
| Input X_PAD\[logical\] (C) | **`32[3]`** | **`32[3]`** |
| Input layout / memory | TILE / DRAM | TILE / DRAM |
| Actual DRAM BW | 23.5 GB/s (8.2% of peak) | 23.8 GB/s (8.3% of peak) |
| Useful BW | 2.2 GB/s (0.8% of peak) | 2.2 GB/s (0.8% of peak) |
| **Slowdown vs PM IDEAL** | **11.8×** | **11.6×** |

`PM_BANDWIDTH = PM_IDEAL` confirms the operation is **100% bandwidth-bound** — compute
contributes 0.09% of the ideal time. The required input bandwidth of **277 GB/s** is near
the Wormhole N150 DRAM peak of ~288 GB/s, yet 90.6% of every read is zero-padding.

For detailed explanation of the compilation chain (ONNX → TTIR → TTNN IR), tile padding
calculation, permute and matmul bottleneck analysis, and profiling results across all four
test cases, see:
[tt_metal_conv2d_dram_bottleneck_detailed.docx](https://github.com/user-attachments/files/28932343/tt_metal_conv2d_dram_bottleneck_detailed.docx)

## Reproduction

```python
# tests/ttnn/unit_tests/operations/conv/test_conv2d_pointwise.py

import pytest
import torch
import ttnn

_CONFIGS = [
    pytest.param(1, 3, 3, 1536, 1536, id="conv2d_1_1x3x1536x1536"),
    pytest.param(1, 3, 3, 1280, 2304, id="conv2d_2_1x3x1280x2304"),
]


def _make_compute_config(device):
    return ttnn.init_device_compute_kernel_config(
        device.arch(),
        math_fidelity=ttnn.MathFidelity.HiFi3,
        fp32_dest_acc_en=True,
        math_approx_mode=True,
    )


def _make_conv_config():
    return ttnn.Conv2dConfig(
        weights_dtype=ttnn.bfloat16,
        output_layout=ttnn.TILE_LAYOUT,
        deallocate_activation=True,
        act_block_h_override=0,
        enable_kernel_stride_folding=False,
        config_tensors_in_dram=True,
    )


@pytest.mark.parametrize(
    "batch, in_channels, out_channels, input_height, input_width",
    _CONFIGS,
)
def test_conv2d_only(device, batch, in_channels, out_channels, input_height, input_width):
    """Isolates the ttnn.conv2d MatmulDeviceOperation kernel.
    Activation pre-placed in DRAM as [1, 1, H*W, in_channels] TILE DRAM —
    exactly as it arrives at conv2d after the compiler-inserted permute+reshape."""
    compute_config   = _make_compute_config(device)
    conv_config      = _make_conv_config()
    spatial          = input_height * input_width
    dram_interleaved = ttnn.DRAM_MEMORY_CONFIG

    tt_weight = ttnn.prepare_conv_weights(
        weight_tensor=ttnn.from_torch(
            torch.randn(out_channels, in_channels, 1, 1, dtype=torch.bfloat16),
            dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT),
        input_memory_config=dram_interleaved, input_layout=ttnn.TILE_LAYOUT,
        weights_format="OIHW", in_channels=in_channels, out_channels=out_channels,
        batch_size=batch, input_height=input_height, input_width=input_width,
        kernel_size=(1, 1), stride=(1, 1), padding=(0, 0, 0, 0), dilation=(1, 1),
        has_bias=True, groups=1, device=device,
        input_dtype=ttnn.bfloat16, output_dtype=ttnn.bfloat16,
        conv_config=conv_config, compute_config=compute_config,
        slice_config=ttnn.Conv2dL1FullSliceConfig,
    )
    tt_bias = ttnn.prepare_conv_bias(
        bias_tensor=ttnn.from_torch(
            torch.randn(1, 1, 1, out_channels, dtype=torch.bfloat16),
            dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT),
        input_memory_config=dram_interleaved, input_layout=ttnn.TILE_LAYOUT,
        in_channels=in_channels, out_channels=out_channels,
        batch_size=batch, input_height=input_height, input_width=input_width,
        kernel_size=(1, 1), stride=(1, 1), padding=(0, 0, 0, 0), dilation=(1, 1),
        groups=1, device=device, input_dtype=ttnn.bfloat16, output_dtype=ttnn.bfloat16,
        conv_config=conv_config, compute_config=compute_config,
        slice_config=ttnn.Conv2dL1FullSliceConfig,
    )
    tt_input = ttnn.from_torch(
        torch.randn(batch, 1, spatial, in_channels, dtype=torch.bfloat16),
        dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
        device=device, memory_config=dram_interleaved,
    )

    [tt_out, [out_h, out_w], [d_w, d_b]] = ttnn.conv2d(
        input_tensor=tt_input, weight_tensor=tt_weight,
        in_channels=in_channels, out_channels=out_channels,
        device=device, bias_tensor=tt_bias,
        kernel_size=(1, 1), stride=(1, 1), padding=(0, 0, 0, 0), dilation=(1, 1),
        batch_size=batch, input_height=input_height, input_width=input_width,
        groups=1, dtype=ttnn.bfloat16,
        conv_config=conv_config, compute_config=compute_config,
        slice_config=ttnn.Conv2dL1FullSliceConfig,
        return_output_dim=True, return_weights_and_bias=True,
    )
    ttnn.deallocate(tt_input)
    ttnn.deallocate(tt_weight)
    ttnn.deallocate(tt_bias)

    result = ttnn.to_torch(tt_out)
    assert result.shape == torch.Size((batch, 1, spatial, out_channels))
    ttnn.deallocate(tt_out)
```

## Request for Inputs

`ttnn.conv2d` with `in_channels=3` and `kernel_size=(1,1)` runs as a `MatmulDeviceOperation`
that is **11.6–11.8× slower than the performance model ideal** on Wormhole N150, with
FPU utilisation of 0.008% and required DRAM bandwidth of 277 GB/s. The root cause is the
32×32 tile padding on the channel dimension after the NHWC permute — 90.6% of every DRAM
read is zero-padding that the kernel immediately discards.

**What is the recommended approach to improve the performance of `ttnn.conv2d`
when `in_channels` is smaller than the tile width (e.g. `in_channels=3`) and the
activation tensor is in DRAM on Wormhole?**

**Comment by chandrasekaranpradeep** [comment]
## Attachments

Four Tracy ops perf CSV reports are attached as [conv2d_dram_bottleneck_reports.zip](https://github.com/user-attachments/files/28872931/conv2d_dram_bottleneck_reports.zip)

**Comment by chandrasekaranpradeep** [comment]
cc @nvukobratTT

**Comment by peter941221** [comment]
I took a closer look at the current conv2d lowering and I think there are three discussion-worthy directions here, with different confidence levels.

The narrowest short-term experiment is to challenge the current 1x1 -> matmul lowering for this pathological shape family.

The current predicate in use_matmul_for_1x1_conv(...) routes 1x1 / stride=1 / pad=0 / dilation=1 into the matmul path by default:
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp#L524-L535

conv2d.cpp then explicitly branches between ttnn::prim::conv2d(...) and ttnn::linear(...):
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp#L257-L367

What makes this worth testing is that the regular conv path is not using the exact same activation contract as the matmul lowering. The conv path explicitly expects row-major activations in the public docs and in the device-side factories:
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_nanobind.cpp#L371-L375
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_sharded_program_factory.cpp#L211-L212
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/device/conv2d_op_width_sharded_program_factory.cpp#L80-L81

Its input-channel alignment rule is also looser than the default tile-width case: for non-mm conv in row-major or sliced form, get_input_channels_alignment(...) can use 8 / 16 / 32 instead of falling straight to 32:
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp#L92-L123

I do not think source inspection is enough to claim that a regular-conv fallback would fix this outright, because conv2d still uses NHWC semantics and still computes padded channel boundaries on the regular-conv side as well:
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp#L152-L159
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp#L1019-L1021

But it does look strong enough to justify a targeted A/B experiment instead of assuming the current 1x1 -> matmul choice is optimal for this shape.

A second, more structural direction is to treat this as a small-C representation problem rather than only a matmul scheduling problem. The codebase already has prior art for channel-alignment-aware representation changes: the kernel-stride-folding docs explicitly call out "small channel counts like 3 RGB channels" as a beneficial case:
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_nanobind.cpp#L429-L441

The folding path already rewrites the effective representation by changing channels / stride / kernel / padding before lowering:
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp#L1384-L1402
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp#L1467-L1470

That makes a small-C-specific path that avoids materializing the 3 -> 32 padded NHWC representation look consistent with existing optimization style, even if it is clearly a broader change.

A third, higher-level direction is to avoid materializing the inflated NHWC intermediate in DRAM at all. The current conv path explicitly flattens to [1, 1, NHW, C] before sharding:
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp#L833-L838

The DRAM sliced path later reshapes back to [N, H, W, C], runs the sliced conv, and then flattens the output again:
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp#L820-L870

More generally, the repo already accepts replacing explicit permute/reshape-style chains with a more local representation rewrite. The fold tests use a reshape -> permute -> reshape reference:
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/tests/ttnn/unit_tests/operations/conv/data_movement/test_fold_op.py#L75-L77

The production fold path uses reshape -> halo -> reshape:
https://github.com/tenstorrent/tt-metal/blob/97ca6204f5aa5d6dbee6fe39da6bd468e4ef42d7/ttnn/cpp/ttnn/operations/data_movement/fold/fold.cpp#L327-L343

That seems like a reasonable precedent for asking whether this lowering can keep the inflated NHWC intermediate from landing in DRAM.

So my current view is:

1. Short term: test whether this pathological 1x1 small-C DRAM case should opt out of the current matmul lowering.
2. Longer term: consider a small-C path that attacks the 3 -> 32 materialization directly.
3. Higher level: check whether the inflated NHWC intermediate can be avoided or kept more local instead of being materialized in DRAM.

**Comment by skhatriTT** [comment]
Hey @fahmedchanTT could you please triage this issue? Thanks!
